src/smt/theory_str_noodler/decision_procedure.h

```diff
@@ -0,0 +1,150 @@
+            auto is_deleted = [&deleted_nodes](std::shared_ptr<GraphNode> node) { return (deleted_nodes.count(node) > 0) ; };
+            nodes_to_process.erase(std::remove_if(nodes_to_process.begin(), nodes_to_process.end(), is_deleted), nodes_to_process.end());
+        }
+
+        /**
+         * @brief Combines aut_ass and substitution_map into one AutAssigment
+         * 
+         * For example, if we have aut_ass[x] = aut1, aut_ass[y] = aut2, and substitution_map[z] = xy, then this will return
+         * automata assignment ret_ass where ret_ass[x] = aut1, ret_ass[y] = aut2, and ret_ass[z] = concatenation(aut1, aut2)
+         * 
+         */
+        AutAssignment flatten_substition_map() {
+            AutAssignment result = aut_ass;
+            std::function<std::shared_ptr<Mata::Nfa::Nfa>(const BasicTerm&)> flatten_var;
+
+            flatten_var = [&result, &flatten_var, this](const BasicTerm &var) -> std::shared_ptr<Mata::Nfa::Nfa> {
+                if (result.count(var) == 0) {
+                    std::shared_ptr<Mata::Nfa::Nfa> var_aut = std::make_shared<Mata::Nfa::Nfa>();
+                    auto state = var_aut->add_state();
+                    var_aut->initial.add(state);
+                    var_aut->final.add(state);
+                    for (const auto &subst_var : this->substitution_map.at(var)) {
+                        var_aut = std::make_shared<Mata::Nfa::Nfa>(Mata::Nfa::concatenate(*var_aut, *flatten_var(subst_var)));
+                    }
+                    result[var] = var_aut;
+                    return var_aut;
+                } else {
+                    return result[var];
+                }
+            };
+            for (const auto &subst_map_pair : substitution_map) {
+                flatten_var(subst_map_pair.first);
+            }
+            return result;
+        }
+    };
+
+    class DecisionProcedure : public AbstractDecisionProcedure {
+    private:
+        // prefix of newly created vars during the procedure
+        const std::string VAR_PREFIX = "tmp";
+        // counter of noodlifications, so that newly created variables will have unique names per noodlification
+        // by for example setting the name to VAR_PREFIX + "_" + noodlification_no + "_" + index_in_the_noodle
+        unsigned noodlification_no = 0;
+
+        std::deque<SolvingState> worklist;
+    public:
+
+
+        DecisionProcedure(const Formula &equalities, AutAssignment init_aut_ass, const std::unordered_set<BasicTerm> init_length_sensitive_vars);
+
+        // returns true if there is something in worklist that is satisfiable and saves the satisfying element in solution
+        bool compute_next_solution();
+        SolvingState solution;
+
+        void initialize(const Instance& inst) override {
+            assert(false); // TODO implement me
+        }
+        bool get_another_solution(const Instance& inst, LengthConstr& out) override {
+            assert(false); // TODO implement me
+        }
```

Comment:
These two functions should no longer exist in the `AbstractDecisionProcedure` neither in the `DecisionProcedure`.